daemon/containerd/image_list.go

```diff
@@ -237,6 +238,77 @@ func (i *ImageService) imageSummary(ctx context.Context, img images.Image, platf
 			return nil
 		}
 
+		mfstSummary := imagetypes.ImageManifestSummary{
+			ID:         target.Digest.String(),
+			Available:  available,
+			Descriptor: target,
+			Kind:       imagetypes.ImageManifestKindUnknown,
+		}
+
+		if opts.Manifests {
+			defer func() {
+				// If the platform is available, prepend it to the list of platforms
+				// otherwise append it at the end.
+				if available {
+					manifestSummaries = append([]imagetypes.ImageManifestSummary{mfstSummary}, manifestSummaries...)
+				} else {
+					manifestSummaries = append(manifestSummaries, mfstSummary)
+				}
+			}()
+		}
+
+		contentSize, err := img.Size(ctx)
+		if err != nil {
+			if !cerrdefs.IsNotFound(err) {
+				log.G(ctx).WithFields(log.Fields{
+					"error":  err,
+					"image":  img.Name(),
+					"digest": target.Digest,
+				}).Warn("failed to determine size")
+			}
+		} else {
+			mfstSummary.Size.Content = contentSize
+			totalSize += contentSize
+			mfstSummary.Size.Total = totalSize
+		}
+
+		isPseudo, err := img.IsPseudoImage(ctx)
+		if err != nil && !errdefs.IsNotFound(err) {
+			log.G(ctx).WithFields(log.Fields{
+				"error":    err,
+				"image":    img.Name(),
+				"digest":   target.Digest,
+				"isPseudo": isPseudo,
+			}).Debug("pseudo image check failed")
+			return nil
+		}
+		if isPseudo {
+			if img.IsAttestation() {
```

Comment:
Perhaps not for this PR, but perhaps some of these may be good to extract to a separate function at some point. I think it may make the flow clearer as there's some early returns happening that are easy to miss because they're a bit buried in the code to construct the types.

Basically thinking along the lines of

1. determine the `Kind`
2. calling function for `Kind` to propagate the relevant data

Reply:
Yep, it's a bit complex at this point, but splitting it into functions isn't great either as there's a lot of shared state (`totalSize`, allChainsIDs`, `containersCount`, `best`).

I can extract it to a separate struct in a follow up (or later, after we decide to settle for this approach).

Reply:
Yup indeed. Perhaps there's not better solution, and definitely OK with me to look at in follow-ups

Reply:
I'll definitely try to improve it, but in this case we might need to sacrifice _some_ readability over performance because it's a hot path 😞